File: RB/ts/RiotBase/Db/Core/DbItemPtr.ts
import Method from "../../Variable/Core/Method";
import Rp     from "../../Variable/Core/Rp";
import UsrId  from "../../System/UsrId";
import DbItem from "./DbItem";

//
export default
class DbItemPtr extends Rp {
    constructor( db_item : DbItem ) {
        super( DbItemPtr );

        this.db_item = db_item
    }

    prox_ptr() {
        return this.db_item.final_item.patch_manager.variable;
    }

    prox_ref() {
        return this.db_item.final_item.patch_manager.variable;
    }

    db_item: DbItem;
}
Rp.make_templ( DbItemPtr );

Method.methods["self_ot_merge_o"].add_surdef( 4, [ DbItemPtr, DbItemPtr ], function( a: { rp: DbItemPtr }, b: { rp: DbItemPtr }, as_usr : UsrId ) : boolean {
    return a.rp.db_item.patch_manager.merge( b.rp.db_item.patch_manager, b.rp.db_item.inode, as_usr );
} );

File: RB/cpp/RiotBase/Db/Core/DbItemPtr.cpp
#include "../../System/Ccast.h"
#include "DbItemPtr.h"
#include "DbItem.h"

namespace RiotBase {

DbItemPtr::DbItemPtr( RcPtr<DbItem> db_item ) : db_item( db_item ) {
}

std::string DbItemPtr::type_name() {
    return "DbItem";
}

Rp **DbItemPtr::prox_ptr() {
    return &db_item->patch_manager.rp;
}

const Rp *DbItemPtr::prox_ref() const {
    return ccast( db_item.ptr() )->patch_manager.rp;
}

void DbItemPtr::write_structure( std::ostream &os ) const {
    os << "DbItemPtr(";
    if ( db_item->patch_manager.rp )
        RO_CALL( write_structure, db_item->patch_manager.rp, os );
    os << ')';
}

void DbItemPtr::self_then( Callback *c ) const {
    RO_CALL( self_then, ccast( db_item.ptr() )->patch_manager.rp, c );
}

} // namespace RiotBase
